test/hotspot/gtest/memory/test_arena.cpp

```diff
@@ -25,9 +25,307 @@
 
 #include "precompiled.hpp"
 #include "memory/arena.hpp"
+#include "runtime/os.hpp"
 #include "utilities/align.hpp"
 #include "utilities/globalDefinitions.hpp"
 #include "unittest.hpp"
+#include "testutils.hpp"
+
+#define ASSERT_CONTAINS(ar, p) ASSERT_TRUE(ar.contains(p))
+
+// Note:
+// - Amalloc returns 64bit aligned pointer (also on 32-bit)
+// - AmallocWords returns word-aligned pointer
+#define ASSERT_ALIGN_AMALLOC(p)       ASSERT_ALIGN(p, ARENA_AMALLOC_ALIGNMENT)
+#define ASSERT_ALIGN_AMALLOCWORDS(p)  ASSERT_ALIGN(p, BytesPerWords)
+
+// Do a couple of checks on the return of a successful Amalloc
+#define ASSERT_AMALLOC(ar, p) \
+  ASSERT_NOT_NULL(p); \
+  ASSERT_CONTAINS(ar, p); \
+  ASSERT_ALIGN_AMALLOC(p);
+
+// #define LOG(s) tty->print_cr s;
+#define LOG(s)
+
+// Test behavior for Amalloc(0).
+// Amalloc just ignores Amalloc(0) and returns the current hwm without increasing it.
+// Therefore, the returned pointer should be not null, aligned, but not (!) contained
+// in the arena since hwm points beyond the arena.
+TEST_VM(Arena, alloc_size_0) {
+  Arena ar(mtTest);
+  void* p = ar.Amalloc(0);
+  ASSERT_NOT_NULL(p);
+  ASSERT_ALIGN_AMALLOC(p);
+  if (!UseMallocOnly) {
+    // contains works differently for malloced mode (and there its broken anyway)
+    ASSERT_FALSE(ar.contains(p));
+  }
+  // Allocate again. The new allocations should have the same position as the 0-sized
+  // first one.
```

Comment:
It seems so dangerous to allow zero sized Amalloc.

Reply:
I agree, but I wanted these tests to test existing behavior.
I tried to do a quick "lets assert > 0" and that backfired, a bit of cleanup would be needed at the caller side.